index/segment/mem/searcher.go

```diff
@@ -0,0 +1,176 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+)
+
+var (
+	errFirstFilterMustNotBeNegation = fmt.Errorf("first filter must be non-negation")
+)
+
+type negationMergeFn func(x segment.PostingsList, y segment.ImmutablePostingsList)
+
+func differenceNegationFn(x segment.PostingsList, y segment.ImmutablePostingsList) { x.Difference(y) }
+func unionNegationFn(x segment.PostingsList, y segment.ImmutablePostingsList)      { x.Union(y) }
+
+type sequentialSearcher struct {
+	queryable       queryable
+	negationMergeFn negationMergeFn
+	pool            segment.PostingsListPool
+}
+
+// newSequentialSearcher returns a new sequential searcher.
+func newSequentialSearcher(
+	q queryable,
+	negationMergeFn negationMergeFn,
+	pool segment.PostingsListPool,
+) searcher {
+	return &sequentialSearcher{
+		queryable:       q,
+		negationMergeFn: negationMergeFn,
+		pool:            pool,
+	}
+}
+
+func (s *sequentialSearcher) Query(query segment.Query) (
+	candidateDocIDs segment.PostingsList,
+	pendingFilterFn matchPredicate,
+	err error,
+) {
+	if err := validateQuery(query); err != nil {
+		return nil, nil, err
+	}
+
+	// order filters to ensure the first filter has no-negation
+	filters := orderFiltersByNonNegated(query.Filters)
+	sort.Sort(filters)
+
+	var (
+		candidateDocIds segment.PostingsList
+		predicates      = make([]matchPredicate, 0, len(query.Filters))
+	)
+	// TODO: support parallel fetching across segments/filters
+	for filterIdx, filter := range query.Filters {
+		if filterIdx == 0 && filter.Negate {
+			return nil, nil, errFirstFilterMustNotBeNegation
+		}
+
+		fetchedIds, pred, err := s.queryable.Filter(filter)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if pred != nil {
+			predicates = append(predicates, pred)
+		}
+
+		// i.e. we don't have any documents for the given filter, can early terminate entire fn
+		if fetchedIds == nil {
+			return nil, nil, nil
+		}
+
+		if candidateDocIds == nil {
+			candidateDocIds = fetchedIds
+			continue
+		}
+
+		// TODO: evaluate perf impact of retrieving all candidate docIDs, waiting till end,
+		// sorting by size and then doing the intersection
+
+		// update candidate set
+		if filter.Negate {
+			s.negationMergeFn(candidateDocIds, fetchedIds)
+		} else {
+			candidateDocIds.Intersect(fetchedIds)
+		}
+
+		// early terminate if we don't have any docs in candidate set
+		if candidateDocIds.IsEmpty() {
+			return nil, nil, nil
+		}
+	}
+
+	// TODO: once we support multiple segments, we'll have to merge results
+	return candidateDocIds, matchPredicates(predicates).Fn(), nil
```

Comment:
Instead of creating a heap allocated slice and a heap allocated fn, wouldn't it be better to return just a single wrapper type that is stack allocated?

i.e.